source/math/FloatVectorList.ooc

```diff
@@ -381,4 +382,68 @@ FloatVectorList: class extends VectorList<Float> {
 			result add(this[i] clamp(floor, ceiling))
 		result
 	}
+	findMaxIndex: func -> Int {
+		index := 0
+		thisPointer := (this pointer as Float*)
+		for (i in 1 .. this count)
+			if (thisPointer[index] < thisPointer[i])
+				index = i
+		index
+	}
+	crossCorrelation: func (other: This, vectorLength: Int) -> Float {
+		upperPart := 0.0f
+		lowerPart1 := 0.0f
+		lowerPart2 := 0.0f
+		thisMean := this mean
+		otherMean := other mean
+		thisPointer := (this pointer as Float*)
+		otherPointer := (other pointer as Float*)
+		for (i in 0 .. vectorLength) {
+			thisValue := thisPointer[i]
+			otherValue := otherPointer[i]
+			thisDiff := (thisValue - thisMean)
+			otherDiff := (otherValue - otherMean)
+			upperPart += thisDiff * otherDiff
+			lowerPart1 += thisDiff * thisDiff
+			lowerPart2 += otherDiff * otherDiff
+		}
+		upperPart / sqrt(lowerPart1 * lowerPart2)
+	}
+	shift: func (offset: Int, filledValue := 0.0f) -> This {
+		result := This new(this count)
+		thisPointer := (this pointer as Float*)
+		for (i in 0 .. this count) {
+			index := i + offset
+			value := (index >= 0 && index < this count) ? thisPointer[index] : filledValue
+			result add(value)
+		}
+		result
+	}
+	// offset < 0 means it shifts to left
+	calculateCorrelation: func (other: This, offsetRange: Range, calculateOffset := true) -> (This, This) {
+		crossCorrelation := This new(offsetRange count)
+		for (offset in offsetRange min .. offsetRange max)
+			crossCorrelation add(this shift(offset) crossCorrelation(other, other count))
+		xValue := This new()
+		xValue add(offsetRange min as Float) . add(offsetRange max as Float)
+		(crossCorrelation, xValue interpolate((offsetRange max - offsetRange min) - 1))
```

Comment:
I would rename `xValue` to `offsets` and you should make the function return `null` for the second tuple value instead of `offsets` if `calculateOffsets` is `false`.